src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
       this.runningThread = null;
       runningThread.interrupt();
     }
-    
-    /**
-     * Called when the queue has been updated and we may need to wake up the consumer thread.
-     */
-    protected void handleQueueUpdate() {
-      if (! startIfNotStarted()) {
-        Thread currRunningThread = runningThread;
-        if (currRunningThread != null) {
-          LockSupport.unpark(currRunningThread);
-        }
-      }
-    }
 
     @Override
     public void run() {
+      Worker worker = null;
       while (runningThread != null) {
         try {
-          TaskWrapper nextTask = getNextTask();
-          if (nextTask != null) {
-            switch (queuePriority) {
-              case High:
-                workerPool.runHighPriorityTask(nextTask);
-                break;
-              case Low:
-                workerPool.runLowPriorityTask(nextTask);
-                break;
-              default:
-                throw new UnsupportedOperationException();
+          worker = workerPool.getWorker();
+          if (worker != null) {
+            TaskWrapper nextTask = getNextReadyTask();
+            if (nextTask != null) {
+              worker.nextTask(nextTask);
+            } else {
+              workerPool.workerDone(worker);
             }
           }
         } catch (InterruptedException e) {
+          if (worker != null) {
+            workerPool.workerDone(worker);
+          }
           stopIfRunning();
           Thread.currentThread().interrupt();
         } catch (Throwable t) {
+          if (worker != null) {
+            // we kill the worker here because this condition is not expected
+            workerPool.killWorker(worker);
+          }
           ExceptionUtils.handleException(t);
+        } finally {
+          worker = null;
         }
       }
     }
     
-    protected TaskWrapper getNextTask() throws InterruptedException {
+    /**
+     * Gets the next task that is ready to execute immediately.  This call gets the next high and 
+     * low priority task, compares them and determines which one should be executed next.
+     * 
+     * If there are no tasks ready to be executed this will block until one becomes available 
+     * and ready to execute.
+     * 
+     * @return Task to be executed next, or {@code null} if shutdown while waiting for task
+     * @throws InterruptedException Thrown if calling thread is interrupted while waiting for a task
+     */
+    protected TaskWrapper getNextReadyTask() throws InterruptedException {
```

Comment:
Take a look at this logic...this is where we decide if we should run the low priority task or the high priority task...the first couple cases are if only one task is in queue...those will be rare.  The bottom case is where we have a task in both queues, and we find out which one is ready to execute, and if both are ready which one should be ran.